src/shared/logs/logs.py

```diff
@@ -149,50 +152,106 @@ def get_raw_events(
-                    ),
-                ),
-                distinct=True,
-            ),
-            package_count=Count("derivation__name", distinct=True),
+
+        maintainer_qs = self._annotate_username(
+            MaintainersEditEvent.objects.prefetch_related(
+                "pgh_context", "maintainer"
+            ).filter(suggestion__in=suggestion_ids)
         )
 
-        for package_event in package_qs.all().iterator():
+        for maintainer_event in maintainer_qs.all().iterator():
             raw_events.append(
-                PackageChangeEvent(
-                    suggestion_id=package_event.proposal_id,
-                    timestamp=package_event.pgh_created_at,
-                    username=package_event.username,
-                    action=package_event.pgh_label,
-                    package_names=package_event.package_names,
-                    package_count=package_event.package_count,
+                MaintainerChangeEvent(
+                    suggestion_id=maintainer_event.suggestion.id,
+                    timestamp=maintainer_event.pgh_created_at,
+                    username=maintainer_event.username,
+                    action=maintainer_event.pgh_label,
+                    # TODO: we should use Pydantic model for maintainers and
+                    # friends as well, at some point
+                    maintainer=cast(
+                        Maintainer, model_to_dict(maintainer_event.maintainer)
+                    ),
                 )
             )
 
         return sorted(raw_events, key=lambda event: event.timestamp)
 
+    def _remove_canceling_events(
+        self,
+        events: list[
+            PackageChangeEvent | SuggestionChangeEvent | MaintainerChangeEvent
+        ],
+        time_threshold_seconds: int = 300,
+    ) -> list[PackageChangeEvent | SuggestionChangeEvent | MaintainerChangeEvent]:
+        """Remove consecutive events that cancel each other out within a time window."""
+
+        filtered_events = []
+        i = 0
+
+        while i < len(events):
+            if i + 1 < len(events) and self._events_cancel_each_other(
+                events[i], events[i + 1], time_threshold_seconds
+            ):
+                # Skip both events
+                i += 2
+            else:
+                # Keep this event
+                filtered_events.append(events[i])
+                i += 1
+
+        return filtered_events
+
+    def _events_cancel_each_other(
+        self,
+        event1: PackageChangeEvent | SuggestionChangeEvent | MaintainerChangeEvent,
+        event2: PackageChangeEvent | SuggestionChangeEvent | MaintainerChangeEvent,
+        time_threshold_seconds: int,
+    ) -> bool:
+        """Check if two consecutive events cancel each other."""
+        # Same user, same suggestion, within time threshold
+        if (
+            event1.username != event2.username
+            or event1.suggestion_id != event2.suggestion_id
```

Comment:
The way we're using it would always give the same suggestion. Is this simply defensive programming?